docs/tokencache.md

```diff
@@ -2,16 +2,350 @@
+
+```java
+//Example
+/**
+     * Saves the credentials and tokens returned by the service to the cache.
+     *
+     * @param oAuth2Strategy The strategy used to create the token request.
+     * @param request        The request used to acquire tokens and credentials.
+     * @param response       The response received from the IdP/STS.
+     * @return The {@link ICacheRecord} containing the Account + Credentials saved to the cache.
+     * @throws ClientException If tokens cannot be successfully saved.
+     */
+    public abstract ICacheRecord save(final T oAuth2Strategy,
+                                      final U request,
+                                      final V response) throws ClientException;
+```
+
+### MSALOAuth2TokenCache
+
+The simplest version of the cache is the MSALOAuth2TokenCache.  There is one instance of this cache per PublicClientApplication created using MSAL within a client android application.  The important thing to note is that there is 1 cache per public client application identified by it's client id.
+
+### Family of Client Ids (Special Case)
+
+Applications published by Microsoft may be authorized to share refresh tokens with one another.  All applications that are authorized to share tokens share a single MSALOAuth2TokenCache instance.  They effectively share a client id, which in this case is the family id (generally 1 is used to represent the family)
+
+### BrokerOAuth2TokenCache
+
+The broker token cache is effectively a wrapper around individiual caches that exist for each client id and the cache that exists for the Microsoft family of clients.  In addition there's a reference to the application metadata cache which tracks which client id/packagename/signature/calling process UI is a member of the Microsoft familiy.  
+
+## Universal Cache Implementation
+
+### AccountCredentialBase
+
+Abstract base class for accounts and credential objects.  The functionality that is common between all objects in the universal schema is the ability to append additional information to both accounts and credentials hence the methods for getting and setting the additional fields are defined here.
+
+### AccountRecord
+
+"Record" is appended in this case to avoid confusion with the public API account object.  This is where you will find all of the information about an account.  This is the object that will be serialized to storage as an Account.
+
+### Credential
+
+This is an abstract base class for all types of credentials stored in the universal cache.  This includes common fields that exist on all types of credentials.  
+
+### AccessTokenRecord
+
+Inherits from Credential and contains additional fields specific to access tokens.
+
+### RefreshTokenRecord
+
+Inherits from Credential and contains additioonal fields specific to refresh tokens.
+
+### PrimaryRefreshTokenRecord
```

Comment:
This is interesting. I didn't actually knew we had this. Do you know what's the purpose of this class? Does having this type of record mean that even PRT is saved in SharedPreferences? From my knowledge, PRT was only saved in Account Manager.

Furthermore, how is this record different from [PRTv2](https://github.com/AzureAD/ad-accounts-for-android/blob/dev/broker4j/src/main/com/microsoft/identity/broker4j/broker/prtv2/PrtV2.java)? I think it would be good to document those differences here (or somewhere).

Reply:
This is being used by OneAuth for a 1P application that authorized to use the PRT protocol.  When we look at the refactoring for the latest PRT protocol.... this is technical issue that we can look into.  @mohitc1 - FYI